### PR TITLE
Add texture

### PR DESCRIPTION
**Pull request by neithernut**
This PR is an extraction of #533.
It provides texture areas.

### CODE REVIEW

**Comment by neithernut** [comment]
Although it's WIP, you may still review it. (comments will be expanded btw.)

File: src/compositor/texture.c [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+    ++level;
+    return self->width >> (level / 2);
+}
+
+uint32_t
+ws_texture_leveled_height(
+    struct ws_texture* self,
+    uint16_t level
+) {
+    return self->width >> (level / 2);
```
Comment (matthiasbeyer) [inline src/compositor/texture.c]: You're using `self->width` here, the function is called `height`!

**Comment by neithernut** [comment]
I'll throw out the texture area stuff now, this PR will only introduce textures.

File: src/compositor/buffer/buffer.h [hunk 2]
```diff
@@ -196,6 +199,23 @@ __ws_nonnull__(1)
 ;
 
 /**
+ * Blit the buffer on a texture
+ *
+ * @warning to not pass NULL to this function! It will crash!
```
Comment (matthiasbeyer) [inline src/compositor/buffer/buffer.h]: Typo. You mean "do not pass"

File: src/compositor/buffer/raw_buffer.c [hunk 3]
```diff
@@ -156,3 +167,25 @@ get_format(
...
+    struct ws_raw_buffer* self = (struct ws_raw_buffer*) self_;
+
+    // bind texture
+    ws_texture_bind(texture, GL_TEXTURE_2D);
+
+    // perform the final update
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, self->stride, self->height, 0,
+                 self->fmt->egl.fmt, self->fmt->egl.type, data);
+
+    return glGetError() == GL_NO_ERROR? 0 : -1;
```
Comment (matthiasbeyer) [inline src/compositor/buffer/raw_buffer.c]: Please add a space before the `?`.

Could this be shorter this way:

```
return -(glGetError() == GL_NO_ERROR);
```

?

File: src/compositor/buffer/buffer.h (hunk 2 above)
Comment (neithernut) [inline src/compositor/buffer/buffer.h]: Fun fact: I copy pasted it.

File: src/compositor/texture.c [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    int res = ws_object_init(&self->obj);
+    if (res < 0) {
+        return res;
+    }
+    self->obj.id = &WS_OBJECT_TYPE_ID_TEXTURE;
+
+    // get texture
+    glGenTextures(1, &self->texture);
+
+    return glGetError() == GL_NO_ERROR? 0 : -1;
```
Comment (matthiasbeyer) [inline src/compositor/texture.c]: Also here, possibility to shorten this?

File: src/compositor/buffer/raw_buffer.c (hunk 3 above)
Comment (neithernut) [inline src/compositor/buffer/raw_buffer.c]: Yep, but it wouldn't be any more readable.

**Comment by matthiasbeyer** [comment]
Looks good to me so far, but I'm a bit unhappy with the `transform2texture` name. I'd like to suggest to use `transform_to_texture` or something like that.

**Comment by neithernut** [comment]
Nah.

**Comment by TheNeikos** [comment]
ACK

**Pull request merged**